### PR TITLE
Thread: Correct Common::Barrier implementation

### PR DESCRIPTION
**Pull request by MerryMage**
`Barrier` as currently implemented is a one-shot synchronisation barrier that is unsafe to be reused. Here we make it reusable.

### CODE REVIEW

File: src/common/thread.h [hunk 1]
```diff
@@ -43,55 +42,55 @@ class Event {
...
-    std::condition_variable m_condvar;
-    std::mutex m_mutex;
+    std::condition_variable condvar;
+    std::mutex mutex;
 };
 
 class Barrier {
 public:
-    Barrier(size_t count) : m_count(count), m_waiting(0) {}
+    Barrier(size_t count) : count(count), waiting(0), generation(0) {}
```
Comment (lioncash) [inline src/common/thread.h]: `count(count)` will cause variable shadowing warnings.

File: src/common/thread.h [hunk 2]
```diff
@@ -43,55 +42,55 @@ class Event {
...
 private:
-    std::condition_variable m_condvar;
-    std::mutex m_mutex;
-    const size_t m_count;
-    size_t m_waiting;
+    std::condition_variable condvar;
+    std::mutex mutex;
+    const size_t count;
+    size_t waiting;
+    size_t generation; ///< Incremented once each time the barrier is used
```
Comment (lioncash) [inline src/common/thread.h]: This can likely just be a regular comment, since it's an implementation detail and not a public part of Barrier's API.

**Comment by MerryMage** [comment]
Addressed the above, thanks @lioncash.

**Comment by bunnei** [comment]
LGTM

**Pull request merged**